src/dialect.ml

```diff
@@ -0,0 +1,106 @@
+open! Stdune
+
+module Filter = struct
+  type t =
+    | No_filter
+    | Action of Loc.t * Action_dune_lang.t
+
+  open Dyn.Encoder
+
+  let to_dyn = function
+    | No_filter ->
+        constr "no_filter" []
+    | Action (loc, action) ->
+        constr "action"
+          [ Loc.to_dyn loc
+          ; Dune_lang.to_dyn (Action_dune_lang.encode action)
```

Comment:
Is this the right way to "dyn"-ify an action? By the way, what is the purpose of `Dyn`? Is it used for serialization?

Reply:
That seems right. `dyn` is only used for crash error reports (`Code_error.raise`). It used in tests as well sometimes. Basically, it's a format aimed at Dune developers rather than the user.